src/ucp/core/ucp_ep.c

```diff
@@ -3227,44 +3239,33 @@ ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
 
 int ucp_ep_do_keepalive(ucp_ep_h ep, ucs_time_t now)
 {
-    ucp_worker_h worker = ep->worker;
-    ucp_lane_index_t lane;
+    ucp_worker_h worker       = ep->worker;
+    ucp_lane_index_t lane     = ucp_ep_config(ep)->key.keepalive_lane;
+    ucp_rsc_index_t rsc_index = ucp_ep_get_rsc_index(ep, lane);
     ucs_status_t status;
-    ucp_rsc_index_t rsc_index;
 
     UCP_WORKER_THREAD_CS_CHECK_IS_BLOCKED(ep->worker);
 
     ucs_assert(!(ep->flags & UCP_EP_FLAG_FAILED));
-    ucs_assert(worker->keepalive.lane_map != 0);
+    ucs_assert((rsc_index != UCP_NULL_RESOURCE) ||
+               (lane == ucp_ep_get_cm_lane(ep)));
 
-    ucs_for_each_bit(lane, worker->keepalive.lane_map) {
-        ucs_assert(lane < UCP_MAX_LANES);
-        rsc_index = ucp_ep_get_rsc_index(ep, lane);
-        ucs_assert((rsc_index != UCP_NULL_RESOURCE) ||
-                   (lane == ucp_ep_get_cm_lane(ep)));
+    ucs_trace("ep %p: do keepalive on lane[%d]=%p ep->flags=0x%x", ep, lane,
+              ep->uct_eps[lane], ep->flags);
 
-        ucs_trace("ep %p: do keepalive on lane[%d]=%p ep->flags=0x%x", ep, lane,
-                  ep->uct_eps[lane], ep->flags);
-
-        status = ucp_ep_do_uct_ep_keepalive(ep, ep->uct_eps[lane], rsc_index, 0,
-                                            NULL);
-        if (status == UCS_ERR_NO_RESOURCE) {
-            continue;
-        } else if (status != UCS_OK) {
-            ucs_diag("unexpected return status from doing keepalive(ep=%p, "
-                     "lane[%d]=%p): %s",
-                     ep, lane, ep->uct_eps[lane], ucs_status_string(status));
-        }
-
-        worker->keepalive.lane_map &= ~UCS_BIT(lane);
+    status = ucp_ep_do_uct_ep_keepalive(ep, ep->uct_eps[lane], rsc_index, 0,
+                                        NULL);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        return 0;
     }
 
-    /* Keepalive round is still not finished for this ep - skip ka_last_round
-     * update */
-    if (worker->keepalive.lane_map != 0) {
-        return 0;
+    if (status != UCS_OK) {
+        ucs_diag("unexpected return status from doing keepalive(ep=%p, "
```

Comment:
`"ep %p lane[%d]=%p: failed to send keepalive (%s)", ep, lane, ep->uct_eps[lane], ucs_status_string(status)`